src/batch_request.cpp

```diff
@@ -17,14 +17,14 @@
 #include "batch_request.hpp"
 
 #include "execute_request.hpp"
+#include "external_types.hpp"
 #include "serialization.hpp"
 #include "statement.hpp"
-#include "types.hpp"
 
 
 extern "C" {
 
-CassBatch* cass_batch_new(CassBatchType type) {
+CassBatch* cass_batch_new(CassSession* session, CassBatchType type) {
```

Comment:
@aholmberg That sounds like a good idea. I like the idea of adding a setting to collection for the default protocol encoding. It will always be the latest but can be downgraded.

RE: Deferring the encoding. This is almost (or just) as expensive as re-encoding because the amount of work to decode to re-encode is not that expensive. Only the size of each element needs to be extracted and that doesn't require a full decode of that value i.e. decode_int32() doesn't need to be used. Deferred encoding was the way it was implemented before so I think the slow path will be even noticed.